ui/raidboss/emulator/data/network_log_converter/LineEvent0x15.ts

```diff
@@ -0,0 +1,143 @@
+import LineEvent from './LineEvent';
+import LogRepository from './LogRepository';
+
+export type LineEvent0x15Parts = [
+  string, // 0: ID
+  string, // 1: Timestamp
+  string, // 2: Source ID
+  string, // 3: Source Name
+  string, // 4: Ability ID
+  string, // 5: Ability Name
+  string, // 6: Target ID
+  string, // 7: Target Name
+  string, // 8: Flags
+  string, // 9: Everything here and lower is complicated
+  string, // 10:
+  string, // 11:
+  string, // 12:
+  string, // 13:
+  string, // 14:
+  string, // 15:
+  string, // 16:
+  string, // 17:
+  string, // 18:
+  string, // 19:
+  string, // 20:
+  string, // 21:
+  string, // 22:
+  string, // 23:
+  string, // 24:
+  string, // 25:
+  string, // 26:
+  string, // 27:
+  string, // 28:
+  string, // 29:
+  string, // 30:
+  string, // 31:
+  string, // 32:
+  string, // 33:
+  string, // 34:
+  string, // 35:
+  string, // 36:
+  string, // 37:
+  string, // 38:
+  string, // 39:
+  string, // 40:
+  string, // 41:
+  string, // 42:
+  string, // 43:
+  string, // 44:
+  string, // 45:
+  string, // 46: Checksum
+];
+
+// Ability hit single target event
+export class LineEvent0x15 extends LineEvent {
+  public id: string;
+  public name: string;
+  public abilityId: string;
+  public abilityName: string;
+  public targetId: string;
+  public targetName: string;
+  public flags: string;
+  public damage: number;
+  public targetHp: string;
+  public targetMaxHp: string;
+  public targetMp: string;
+  public targetMaxMp: string;
+  public targetX: string;
+  public targetY: string;
+  public targetZ: string;
+  public targetHeading: string;
+  public sourceHp: string;
+  public sourceMaxHp: string;
+  public sourceMp: string;
+  public sourceMaxMp: string;
+  public x: string;
+  public sourceX: string;
+  public y: string;
+  public sourceY: string;
+  public z: string;
+  public sourceZ: string;
+  public heading: string;
+  public sourceHeading: string;
+
+  constructor(repo: LogRepository, line: string, public parts: LineEvent0x15Parts) {
+    super(repo, line, parts);
+
+    this.id = parts[2].toUpperCase();
+    this.name = parts[3];
+    repo.updateCombatant(this.id, {
+      job: undefined,
+      name: this.name,
+      spawn: this.timestamp,
+      despawn: this.timestamp,
+    });
+
+    this.abilityId = parts[4].toUpperCase();
+    this.abilityName = parts[5];
+
+    this.targetId = parts[6].toUpperCase();
+    this.targetName = parts[7];
+    repo.updateCombatant(this.targetId, {
+      job: undefined,
+      name: this.targetName,
+      spawn: this.timestamp,
+      despawn: this.timestamp,
+    });
+
+    this.flags = parts[8];
+
+    let offset = 0;
+
+    if (this.flags === '3F')
+      offset = 2;
+
+    this.damage = LineEvent.calculateDamage(parts[9 + offset] as string);
+
+    this.targetHp = parts[24 + offset] as string;
```

Comment:
What are all the `as string` doing here? Can they be removed?

Reply:
More weird behavior with `(string | undefined)`. Removing the `as string` for these results in the same errors as outlined in the 0x00 file, despite the `parts` parameter being properly typed, because the indexing of `parts` is being adjusted by the offset.

```
ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x15.ts
116:44-61
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x15.ts(116,45)
      TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x15.ts
118:4-17
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x15.ts(118,5)
      TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
```

Reply:
I think if `parts` is just a `string[]` and you treat everything as potentially undefined (either by `this.hp = parts[7] ?? '0') or `hp?: number` in its declaration) then you don't need `as string` here.